---
title: "Snakemake & Singularity lecture"
author: Sylvain Schmitt
date: January 4th, 2022
output:
  github_document: 
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(kableExtra)
if(knitr:::is_html_output()) options(knitr.table.format = "html") 
if(knitr:::is_latex_output()) options(knitr.table.format = "latex") 
library(tidyverse)
library(vcfR)
library(Biostrings)
theme_set(bayesplot::theme_default())
opts_chunk$set(echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
               cache = F, cache.lazy = F)
```

[`singularity` & `snakemake`](https://github.com/sylvainschmitt/snakemake_singularity) workflow for a lecture at ECOFOG.

```{r dag}
knitr::include_graphics("dag/dag.minimal.svg")
```


# Installation

- [x] Python ≥3.5
- [x] Snakemake ≥5.24.1
- [x] Golang ≥1.15.2
- [x] Singularity ≥3.7.3
- [x] This workflow

```{bash, eval=F, echo=T}
# Python
sudo apt-get install python3.5
# Snakemake
sudo apt install snakemake`
# Golang
export VERSION=1.15.8 OS=linux ARCH=amd64  # change this as you need
wget -O /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz https://dl.google.com/go/go${VERSION}.${OS}-${ARCH}.tar.gz && \
sudo tar -C /usr/local -xzf /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz
echo 'export GOPATH=${HOME}/go' >> ~/.bashrc && \
echo 'export PATH=/usr/local/go/bin:${PATH}:${GOPATH}/bin' >> ~/.bashrc && \
source ~/.bashrc
# Singularity
mkdir -p ${GOPATH}/src/github.com/sylabs && \
  cd ${GOPATH}/src/github.com/sylabs && \
  git clone https://github.com/sylabs/singularity.git && \
  cd singularity
git checkout v3.7.3
cd ${GOPATH}/src/github.com/sylabs/singularity && \
  ./mconfig && \
  cd ./builddir && \
  make && \
  sudo make install
# detect Mutations
git clone git@github.com:sylvainschmitt/detectMutations.git
cd detectMutations
```

# Data

# Usage

## Locally

```{bash, eval=F, echo=T}
snakemake -np -j 3 --resources mem_mb=10000 # dry run
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
snakemake --use-singularity -j 3 --resources mem_mb=10000 # run
```

## HPC

```{bash, eval=F, echo=T}
module load bioinfo/snakemake-5.25.0 # for test on node
snakemake -np # dry run
sbatch job.sh # run
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
```

# Workflow

<!-- ## Reference -->

<!-- *Copy and index reference.* -->

<!-- ### [cp_reference](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/cp_reference.smk) -->

<!--   - Tools: `cp` -->

<!-- ### [bwa_index](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/bwa_index.smk) -->

<!--   - Tools: [`BWA index`](http://bio-bwa.sourceforge.net/bwa.shtml) -->
<!--   - Singularity: -->
<!--     oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest -->

<!-- ## Reads -->

<!-- *Trim and quality check reads.* -->

<!-- ### [trimmomatic](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/trimmomatic.smk) -->

<!-- * Tools: [`Trimmomatic`](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/trimmomatic/trimmomatic:latest -->

<!-- ### [fastqc](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/fastqc.smk) -->

<!-- * Tools: [`fastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) -->
<!-- * Singularity: docker://biocontainers/fastqc:v0.11.9_cv8 -->

<!-- ### [multiqc](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/multiqc.smk) -->

<!-- * Tools: [`MultiQC`](https://multiqc.info/) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/multiqc/multiqc:latest -->

<!-- ## Alignments -->

<!-- *Align reads against reference, mark duplicated, and report alignment quality.* -->

<!-- ### [bwa_mem](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/bwa_mem.smk) -->

<!-- * Tools: [`BWA mem`](http://bio-bwa.sourceforge.net/bwa.shtml) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest -->

<!-- ### [samtools_view](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_view.smk) -->

<!-- * Tools: [`Samtools view`](http://www.htslib.org/doc/samtools-view.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [samtools_sort](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_sort.smk) -->

<!-- * Tools: [`Samtools sort`](http://www.htslib.org/doc/samtools-sort.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [samtools_index](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_index.smk) -->

<!-- * Tools: [`Samtools index`](http://www.htslib.org/doc/samtools-index.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [gatk_markduplicates](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/gatk_markduplicates.smk) -->

<!-- * Tools: [`gatk MarkDuplicates`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-) -->
<!-- * Singularity: docker://broadinstitute/gatk -->

<!-- ### [samtools_view_md](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_view_md.smk) -->

<!-- * Tools: [`Samtools view`](http://www.htslib.org/doc/samtools-view.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [samtools_index_md](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_index_md.smk) -->

<!-- * Tools: [`Samtools index`](http://www.htslib.org/doc/samtools-index.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [samtools_stats](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/samtools_stats.smk) -->

<!-- * Tools: [`Samtools stats`](http://www.htslib.org/doc/samtools-stats.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest -->

<!-- ### [mosdepth](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/mosdepth.smk) -->

<!-- * Tools: [`mosdepth`](https://github.com/brentp/mosdepth) -->
<!-- * Singularity: docker://quay.io/biocontainers/mosdepth:0.2.4--he527e40_0 -->

<!-- ## Mutations -->

<!-- *Detect mutations.* -->

<!-- ### [strelka2](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/strelka2.smk) -->

<!-- * Tools: [`Strelka2`](https://github.com/Illumina/strelka) -->
<!-- * Singularity: docker://quay.io/wtsicgp/strelka2-manta -->

<!-- ### [bedtools_subtract](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/bedtools_subtract.smk) -->

<!-- * Tools: [`bedtools subtract`](https://bedtools.readthedocs.io/en/latest/content/tools/subtract.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bedtools/bedtools:latest -->

<!-- ### [bedtools_intersect](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/bedtools_intersect.smk) -->

<!-- * Tools: [`bedtools intersect`](https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bedtools/bedtools:latest -->

<!-- ### [strelka2tsv](https://github.com/sylvainschmitt/detectMutations/blob/angela/rules/strelka2tsv.smk) -->

<!-- * Script: [`strelka2tsv.R`](https://github.com/sylvainschmitt/detectMutations/blob/angela/scripts/strelka2tsv.R) -->
<!-- * Singularity: to be added, currently uses local install -->



